pkg/api/summary_test.go

```diff
@@ -0,0 +1,46 @@
+package api
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+)
+
+const (
+	// Sample API key for test case usage.
+	testKey = "c808f484a4e88cc32a9a8bfcce19169c77bcd9c5eec18d859e1bb4b318bf42bf"
+)
+
+// Tests for api.Summary.Update() with an API key
+func TestUpdateWithApiKey(t *testing.T) {
+	summary := NewSummary()
+	mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Ensure URL is formatted with the correct query string.
+		if !strings.Contains(r.URL.RequestURI(), "/api.php?summary&auth="+testKey) {
+			t.Error("@TestUpdateWithApiKey: api.Summary.Update() did not request the expected Pi Hole api endpoint with expected API key.")
+		}
+	}))
+	defer mockServer.Close()
+	url := mockServer.URL + "/api.php"
+
+	summary.Update(url, testKey, nil)
+}
+
+// Tests for api.Summary.Update() without an API key
+func TestUpdateWithoutAPIKey(t *testing.T) {
+	summary := NewSummary()
+	mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Ensure URL does not contain &auth as part of its query string.
+		if strings.Contains(r.URL.RequestURI(), "/api.php?summary&auth=") {
+			t.Error("@TestUpdateWithoutAPIKey: api.Summary.Update() did not send the expected query string when calling with an API key: /api.php?summary&auth=")
+		}
+		// Ensure URL is formatted with the correct query string.
+		if !strings.Contains(r.URL.RequestURI(), "/api.php?summary") {
```

Comment:
For constant testing data like `"/api.php?summary"` here, I am going to propose that they are extracted to variables so that they are easier to maintain, as I can see it used in more than one place in this test, as well as potentially in others.

However, as there are multiple tests that will end using strings like this (e.g. the API's root address etc...) do you think it would be worth creating a new testing utils file with some consts defined just so string defs are a bit less repetitive?

Reply:
^ building on this, when working in other codebases that end up growing to include tests, I've found it quite useful to spin off new packages to contain all the tests so that they are not so interspersed among what they are testing. This makes it much clearer where and what is test code versus actual program code.

So something like:

```
.
└── pkg/
    ├── api/
    │   └── ...
    ├── auth/
    │   └── ...
    ├── cli/
    │   └── ...
    ├── data/
    │   └── ...
    ├── database/
    │   └── ...
    ├── network/
    │   └── ...
    ├── ui/
    │   └── ...
    └── test/
        ├── api/
        │   └── summary_test.go
        ├── auth/
        │   └── auth_test.go
        ├── cli
        ├── data
        ├── database
        ├── network
        ├── ui
        └── utils/
            ├── test_constants.go
            └── ( ...or whatever else :) )
```

However if you want to spin this off into another issue along with other larger refactoring you may want to do, then I'm also happy with that.

Reply:
> For constant testing data like `"/api.php?summary"` here, I am going to propose that they are extracted to variables so that they are easier to maintain, as I can see it used in more than one place in this test, as well as potentially in others.
> 
> However, as there are multiple tests that will end using strings like this (e.g. the API's root address etc...) do you think it would be worth creating a new testing utils file with some consts defined just so string defs are a bit less repetitive?

I agree with this. Especially as we move forward with version 6 of the web gui that is set to come out sometime in the future with a brand new API, which will most likely see us making big changes to this entire project (https://discourse.pi-hole.net/t/is-there-an-http-api-documentation/56113/2)

It would be nice to have these test utilities defined in a single file that is accessed as needed throughout our tests.

> ^ building on this, when working in other codebases that end up growing to include tests, I've found it quite useful to spin off new packages to contain all the tests so that they are not so interspersed among what they are testing. This makes it much clearer to see where/what is test code versus what isn't.
> 
> So something like:
> 
> ```
> .
> └── pkg/
>     ├── api/
>     │   └── ...
>     ├── auth/
>     │   └── ...
>     ├── cli/
>     │   └── ...
>     ├── data/
>     │   └── ...
>     ├── database/
>     │   └── ...
>     ├── network/
>     │   └── ...
>     ├── ui/
>     │   └── ...
>     └── test/
>         ├── api/
>         │   └── summary_test.go
>         ├── auth/
>         │   └── auth_test.go
>         ├── cli
>         ├── data
>         ├── database
>         ├── network
>         ├── ui
>         └── utils/
>             ├── test_constants.go
>             └── ( ...or whatever else :) )
> ```
> 
> However if you want to spin this off into another issue along with other larger refactoring you may want to do, then I'm also happy with that.

I also agree with this, and I always prefer to have a dedicated directory to store tests in. Go, however, likes it when tests are part of the same package that is being tested, and it is useful for packages that have variables and functions we need to test that are not exported. Regardless of the structure for where we store our tests that Go expects, it does not forbid a dedicated directory to store tests in.

This is a good article that explains Go conventions for testing: https://ieftimov.com/posts/testing-in-go-naming-conventions/

Let me know your thoughts on this.

Reply:
I didn't realise that Go tests were quite so pedantic 😂 But okay, I see where you're coming from. If this is custom for Go projects then I'm likely just trying to force my own structural ideals onto it without any real reason to. Plus, as that article mentioned, if the `go test` command specifically looks for the `_test.go` suffix, then that is technically a differentiator between tests and and normal code that is enforced by the toolchain. I always get the heebie-jeebies about developer-imposed naming schemes that are not universally enforced; but if the toolchain makes this check for us, then I am happy.

This looks good to merge in this case.